File: hackerrank/problem solving/algorithms/2 - implementation/51 - service lane/service_lane.go
// https://www.hackerrank.com/challenges/service-lane/problem?isFullScreen=true

package main

import (
    "fmt"
    "slices"
)

func main() {
    var n, t int
    _, _ = fmt.Scan(&n, &t)
    widths := readNumbers(n)
    for range t {
        fmt.Println(minWidthInSegment(widths, readNumbers(2)))
    }
}

func readNumbers(n int) []int {
    arr := make([]int, n)
    for i := range n {
        _, _ = fmt.Scan(&arr[i])
    }
    return arr
}

// n: length of widths
// T: O(n)
// S: O(1) extra space
func minWidthInSegment(widths, segment []int) int {
    start, finish := segment[0], segment[1]
    return slices.Min(widths[start : finish + 1])
}

File: hackerrank/problem solving/algorithms/2 - implementation/51 - service lane/service_lane.c
// https://www.hackerrank.com/challenges/service-lane/problem?isFullScreen=true

#include <stdio.h>

typedef struct {
    int entry;
    int exit;
} segment_pair;

void read_numbers(int *arr, int n);
int min_width_in_segment(const int *widths, const segment_pair *segment);

int main()
{
    int n, t;
    scanf("%d %d", &n, &t);
    int widths[n];
    read_numbers(widths, n);

    for (int i = 0; i < t; ++i) {
        segment_pair segment;
        scanf("%d %d", &segment.entry, &segment.exit);
        printf("%d\n", min_width_in_segment(widths, &segment));
    }

    return 0;
}

void read_numbers(int *arr, int n)
{
    for (int i = 0; i < n; ++i)
        scanf("%d", &arr[i]);
}

// n: length of array widths
// T: O(n)
// S: O(1) extra space
int min_width_in_segment(const int *widths, const segment_pair *segment)
{
    int min = widths[segment->entry];
    for (int i = segment->entry; i <= segment->exit; ++i)
        min = min < widths[i] ? min : widths[i];
    return min;
}
